torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -238,15 +262,58 @@ def to_sharded(self) -> None:
         self.free_all_gather_output()
         self.sharded_state = ShardedState.SHARDED
 
+    def to_sharded_post_forward(self) -> None:
+        if self.is_dtensor:
+            raise NotImplementedError(
+                "Resharding to smaller mesh with TP is not supported yet"
+            )
+        self._assert_in_states(ShardedState.UNSHARDED)
+        assert self.post_forward_mesh_info is not None  # mypy
+        shard_world_size = self.post_forward_mesh_info.shard_mesh_size
+        shard_rank = self.post_forward_mesh_info.shard_mesh_rank
+        if (numel := self.all_gather_output.numel()) % shard_world_size != 0:
+            _raise_assert_with_print(
+                f"All-gather output size ({self.all_gather_output.numel()}) must "
+                f"be divisible by the shard world size ({shard_world_size})"
+            )
+        self._sharded_post_forward_param_data = torch.as_strided(
+            self.all_gather_output,
```

Comment:
Since `self.all_gather_output` is 1D, we can replace this `as_strided` with a slice to simplify.
```
self.all_gather_output.narrow(0, numel // shard_world_size * shard_rank, numel // shard_world_size)
```